src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp

```diff
@@ -26,13 +26,206 @@
 namespace ArborX::Interpolation::Details
 {
 
+template <typename CRBFunc, typename PolynomialDegree, typename SourcePoints,
+          typename TargetAccess, typename Coefficients, typename ExecutionSpace>
+class MovingLeastSquaresCoefficientsKernel
+{
+private:
+  using ScratchMemorySpace = typename ExecutionSpace::scratch_memory_space;
+
+  using SourcePoint = typename SourcePoints::non_const_value_type;
+  using TargetPoint = typename TargetAccess::value_type;
+
+  using CoefficientsType = typename Coefficients::non_const_value_type;
+
+  static constexpr int dimension = GeometryTraits::dimension_v<SourcePoint>;
+  static constexpr int degree = PolynomialDegree::value;
+  static constexpr int poly_size = polynomialBasisSize<dimension, degree>();
+
+  template <typename T>
+  using UnmanagedView =
+      Kokkos::View<T, ScratchMemorySpace, Kokkos::MemoryUnmanaged>;
+
+  using LocalSourcePoints = Kokkos::Subview<SourcePoints, int, Kokkos::ALL_t>;
+  using LocalPhi = UnmanagedView<CoefficientsType *>;
+  using LocalVandermonde = UnmanagedView<CoefficientsType **>;
+  using LocalMoment = UnmanagedView<CoefficientsType **>;
+  using LocalSVDDiag = UnmanagedView<CoefficientsType *>;
+  using LocalSVDUnit = UnmanagedView<CoefficientsType **>;
+  using LocalCoefficients = Kokkos::Subview<Coefficients, int, Kokkos::ALL_t>;
+
+public:
+  MovingLeastSquaresCoefficientsKernel(ExecutionSpace const &,
+                                       TargetAccess const &target_access,
+                                       SourcePoints const &source_points,
+                                       Coefficients const &coefficients)
+      : _target_access(target_access)
+      , _source_points(source_points)
+      , _coefficients(coefficients)
+      , _num_targets(target_access.size())
+      , _num_neighbors(source_points.extent_int(1))
+  {}
+
+  std::size_t team_shmem_size(int const team_size) const
+  {
+    std::size_t val = 0;
+    val += LocalPhi::shmem_size(_num_neighbors);
+    val += LocalVandermonde::shmem_size(_num_neighbors, poly_size);
+    val += LocalMoment::shmem_size(poly_size, poly_size);
+    val += LocalSVDDiag::shmem_size(poly_size);
+    val += LocalSVDUnit::shmem_size(poly_size, poly_size);
+    return team_size * val;
+  }
+
+  template <typename TeamMember>
+  KOKKOS_FUNCTION void operator()(TeamMember member) const
+  {
+    int const rank = member.team_rank();
+    int const size = member.team_size();
```

Comment:
Do you really want these local variable?  They are only used to determine the target point index.

Reply:
Or for readability defer to a function that takes the index as argument.

Reply:
They were used elsewhere before, I will remove them.